### DISCUSSION TITLE
Granular field control on GraphQL schemas

### DISCUSSION THREAD

**Discussion by andrewfairlie** [craftcms/cms#9014] [Ideas]
### Description

At the moment a public schema would have access to each of the fields within a channel that the schema has access to. This could, in some conditions, cause security or privacy issues by revealing fields that shouldn't really be visible to anybody without permission.

### Over-the-top example scenario
A website has a Staff Directory, and so there's a "Staff" channel with their names, email addresses, bio in. These are fine to be publically accessible, for example to power a search engine. But there are also additional fields such as Line Manager, Salary, and Annual Review scores. Those fields shouldn't be public.

### The suggestion
Have the ability to pick which fields within a section the schema is allowed to access, and if it's left blank they can access all fields as-per the current functionality.

### The current workaround
Currently, we wouldn't expose the staff channel to the public schema at all and write a middleware API to show only the stuff we're comfortable showing via a private schema. It works, but it's a few extra steps and moves us further away from the neatness of GraphQL.

This kind of power-user tool would set Craft up to comfortable handle some more complex headless websites with ease - but I appreciate this is comfortably in the power-user/niche category.

**Comment by andris-sevcenko** [comment]
While we don't have this implemented, you can use this code to pull this off yourself!

```php
use craft\events\DefineGqlTypeFieldsEvent;
use craft\gql\TypeManager;
use yii\base\Event;

Event::on(TypeManager::class, TypeManager::EVENT_DEFINE_GQL_TYPE_FIELDS, function(DefineGqlTypeFieldsEvent $event) {
    // Remove unneeded fields from the staff entry type in staff section
    if ($event->typeName == 'staff_staff_Entry') {
        unset($event->fields['salary'], $event->fields['review']);
    }
});
```


The granular control you mention will probably happen eventually, but it's not possible at this time to say when.
